### DISCUSSION TITLE
Recursive schema

### DISCUSSION THREAD

**Discussion by macmillen** [streamich/json-joy#690] [Q&A]
How can I convert this type to a json-joy schema?
``` ts
type Block = {
  id: string;
  children?: Block[];
} 
```

I have two problems:
1. how to define a recursive schema
2. how to define optional fields

When I try to do it by just using types without the schema builder:
``` ts
type Block = n.obj<{
  id: n.con<string>;
  children: n.arr<Block[]>;
}>;
```
I get an error:
```
Type of property 'children' circularly references itself in mapped type '{ __id: string; children: any; }'
```

I checked how zod (a TypeScript schema library) is doing it and they do it asynchronously. 
https://github.com/colinhacks/zod#recursive-types

**Comment by streamich** [comment]
Link to zod recursive types: https://github.com/colinhacks/zod/tree/v3?tab=readme-ov-file#recursive-types
